*** qemu-0.12.5/vl.c	2010-07-22 14:39:04.000000000 +0200
--- qemu-0.12.5-patched/vl.c	2010-11-12 13:14:42.000000000 +0100
***************
*** 228,233 ****
--- 228,234 ----
  unsigned int nb_prom_envs = 0;
  const char *prom_envs[MAX_PROM_ENVS];
  int boot_menu;
+ static void e9write(void *opaque, uint32_t address, uint32_t value);

  typedef struct FWBootEntry FWBootEntry;

***************
*** 1453,1458 ****
--- 1454,1460 ----
      }
      bdrv_close_all();
      pause_all_vcpus();
+     e9write(0, 0, '\n');
  }

  static void version(void)
*************** static int virtcon_parse(const char *dev
*** 4848,4853 ****
--- 4848,4891 ----
      return 0;
  }
  
+ static void e9write(void *opaque, uint32_t address, uint32_t value)
+ {
+   static uint32_t last = -1;
+   static size_t pos = 0;
+   static char buffer[256];
+   if (last == -1)
+     buffer[0] = 0;
+   if (value == '\r')
+   {
+     printf("%s\n", buffer);
+     fflush(stdout);
+     pos = 0;
+     buffer[0] = 0;
+   }
+   else if (value == '\n')
+   {
+     if (last != '\r')
+     {
+       printf("%s\n", buffer);
+       fflush(stdout);
+       pos = 0;
+       buffer[0] = 0;
+     }
+   }
+   else
+   {
+     buffer[pos++] = (char) value;
+     buffer[pos] = 0;
+     if (pos+2 >= sizeof(buffer))
+     {
+       printf("%s", buffer);
+       pos = 0;
+       buffer[0] = 0;
+     }
+   }
+   last = value;
+ }
+ 
  int main(int argc, char **argv, char **envp)
  {
      const char *gdbstub_dev = NULL;
*************** int main(int argc, char **argv, char **e
*** 6136,6141 ****
--- 6174,6182 ----
      }
  #endif
  
+     /* register the 0xe9 port to the e9write function */
+     register_ioport_write(0xe9, 1, 1, e9write, NULL);
+ 
      main_loop();
      quit_timers();
      net_cleanup();
